### PR TITLE
util: support futures for mpsc channel

### PR DESCRIPTION
**Pull request by BusyJay**
In the future, peer will be polled as a FSM, hence we need to implement futures support for channels. Note that this is a neither complete nor general implementation, as asynchronously sending is not supported. For loosely bounded channel it's not a problem as sending will never block.

### CODE REVIEW

File: src/util/mpsc.rs [hunk 1]
```diff
@@ -169,6 +190,27 @@ pub fn bounded<T>(cap: usize) -> (Sender<T>, Receiver<T>) {
...
 
+impl<T> Stream for Receiver<T> {
+    type Error = ();
+    type Item = T;
+
+    fn poll(&mut self) -> Poll<Option<T>, ()> {
+        match self.try_recv() {
+            Ok(t) => Ok(Async::Ready(Some(t))),
+            Err(TryRecvError::Empty) => {
+                let t = self.state.recv_task.swap(task::current(), Ordering::AcqRel);
```
Comment (siddontang) [inline src/util/mpsc.rs]: can you use another name instead of `t` to distinguish with above `t`?

**Comment by siddontang** [comment]
PTAL @Hoverbear @overvenus

File: src/util/mpsc.rs [hunk 2]
```diff
@@ -323,6 +367,37 @@ mod tests {
         assert!((i64::from(timer.elapsed().subsec_nanos()) - 100_000_000).abs() < 1_000_000);
     }
 
+    #[test]
+    fn test_notify() {
+        let (tx1, rx1) = super::unbounded();
+        let (tx2, rx2) = mpsc::unbounded();
+        let mut rx2 = rx2.wait();
```
Comment (zhangjinpeng1987) [inline src/util/mpsc.rs]: Will block at here?

Comment (BusyJay) [reply src/util/mpsc.rs]: No, it just
> Creates an iterator which blocks the current thread until each item of this stream is resolved.

**Comment by siddontang** [comment]
LGTM

**Comment by BusyJay** [comment]
@overvenus PTAL

File: src/util/mpsc.rs [hunk 3]
```diff
@@ -93,9 +112,11 @@ impl<T> Sender<T> {
     pub fn try_send(&self, t: T) -> Result<(), TrySendError<T>> {
         if !self.state.is_receiver_closed() {
             crossbeam_channel::select! {
-                send(self.sender, t) => Ok(()),
-                default => Err(TrySendError::Full(t)),
+                send(self.sender, t) => {},
```
Comment (hicqu) [inline src/util/mpsc.rs]: Is it equal to `self.sender.send(t)`?

**Comment by hicqu** [comment]
LGTM.

File: src/util/mpsc.rs (hunk 3 above)
Comment (overvenus) [reply src/util/mpsc.rs]: Why not put `notify` into the send operation body

File: src/util/mpsc.rs [hunk 4]
```diff
@@ -169,6 +190,27 @@ pub fn bounded<T>(cap: usize) -> (Sender<T>, Receiver<T>) {
...
+        match self.try_recv() {
+            Ok(m) => Ok(Async::Ready(Some(m))),
+            Err(TryRecvError::Empty) => {
+                let t = self.state.recv_task.swap(task::current(), Ordering::AcqRel);
+                if t.is_some() {
+                    return Ok(Async::NotReady);
+                }
+                // Note: If there is a message or all the senders are dropped after
+                // polling but before task is set, `t` should be None.
+                self.poll()
```
Comment (overvenus) [inline src/util/mpsc.rs]: Why not replace the recursion with a loop?

File: src/util/mpsc.rs (hunk 3 above)
Comment (BusyJay) [reply src/util/mpsc.rs]: Reduce indent.

File: src/util/mpsc.rs (hunk 4 above)
Comment (BusyJay) [reply src/util/mpsc.rs]: It's easy to understand and there is only one level of recursion.

File: src/util/mpsc.rs (hunk 3 above)
Comment (overvenus) [reply src/util/mpsc.rs]: But it can save an extra notification.

Comment (BusyJay) [reply src/util/mpsc.rs]: I don't get it. What's the extra notification?

Comment (overvenus) [reply src/util/mpsc.rs]: Okay, misread. I thought `Full(t)` also causes a notification.

**Comment by BusyJay** [comment]
PTAL

**Pull request merged**